src/expression.hpp

```diff
@@ -72,37 +77,42 @@ class expression {
 
         ddwaf::timer &deadline;
         const ddwaf::object_limits &limits;
-        const std::vector<std::unique_ptr<condition>> &conditions;
+        const std::vector<condition> &conditions;
         const object_store &store;
-        const std::unordered_set<const ddwaf_object *> &objects_excluded;
+        const exclusion::object_set &objects_excluded;
         const std::unordered_map<std::string, std::shared_ptr<matcher::base>> &dynamic_matchers;
         cache_type &cache;
     };
 
     expression() = default;
 
-    explicit expression(
-        std::vector<std::unique_ptr<condition>> &&conditions, ddwaf::object_limits limits = {})
+    explicit expression(std::vector<condition> &&conditions, ddwaf::object_limits limits = {})
         : limits_(limits), conditions_(std::move(conditions))
     {}
 
-    bool eval(cache_type &cache, const object_store &store,
-        const std::unordered_set<const ddwaf_object *> &objects_excluded,
+    eval_result eval(cache_type &cache, const object_store &store,
+        const exclusion::object_set &objects_excluded,
         const std::unordered_map<std::string, std::shared_ptr<matcher::base>> &dynamic_matchers,
         ddwaf::timer &deadline) const;
 
     void get_addresses(std::unordered_map<target_index, std::string> &addresses) const
     {
         for (const auto &cond : conditions_) {
-            for (const auto &target : cond->targets) {
-                addresses.emplace(target.root, target.name);
-            }
+            for (const auto &target : cond.targets) { addresses.emplace(target.root, target.name); }
         }
     }
 
-    static memory::vector<event::match> &&get_matches(cache_type &cache)
+    static memory::vector<event::match> get_matches(cache_type &cache)
```

Comment:
not sure about this plus the changes in expression::cache_type / expression::evaluator::eval(). Can't the matches be moved? what's the point of keeping the ephemeral condition::cache_type matches? Should they actually use memory that can only be freed at context free, seeing that they refer to data that is valid only for one run?

Reply:
> not sure about this plus the changes in expression::cache_type / expression::evaluator::eval(). Can't the matches be moved?

Ephemeral matches could be moved, I'll check...

> what's the point of keeping the ephemeral condition::cache_type matches?

We don't need to keep them.

> Should they actually use memory that can only be freed at context free, seeing that they refer to data that is valid only for one run?

I guess these could use the standard allocator, I'll think about it.